crepecake/src/main/java/com/google/cloud/tools/crepecake/cache/CacheChecker.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.crepecake.cache;
+
+import com.google.cloud.tools.crepecake.image.ImageLayers;
+import com.google.cloud.tools.crepecake.image.ReferenceLayer;
+import java.io.File;
+import java.io.IOException;
+import java.util.Set;
+
+/** Checks if cached data is outdated. */
+public class CacheChecker {
+
+  private final Cache cache;
+
+  /**
+   * @return the last modified time for the file. Recursively finds the largest last modified time
+   *     for all subfiles if {@code file} is a directory.
+   */
+  private static long getLastModifiedTime(File file) throws IOException {
+    long lastModifiedTime = file.lastModified();
+
+    if (file.isDirectory()) {
```

Comment:
1. What happens if this is a link? If so we could get stuck in an infinite loop. 
2. You probably want file.canRead() rather than isDirectory. There are unreadable directories due to permissions.